File: apps/sdkwork-webserver-flutter-mobile/packages/sdkwork_webserver_flutter_mobile_shell/lib/src/navigation/route_registry.dart
/// Flutter route placement metadata and projection inputs.
///
/// Authority: `FLUTTER_APP_MOBILE_ARCHITECTURE_SPEC.md` and
/// `APP_CLIENT_ARCHITECTURE_ALIGNMENT_SPEC.md` section 7. Route ids follow
/// `<surface>.<domain>.<capability>.<screen>` and stay aligned with the PC, H5,
/// mini program, and HarmonyOS roots; the physical Flutter route name may
/// differ.
///
/// Route metadata must not declare HTTP API paths, SDK methods, raw URL
/// constants, or transport details.
library;

enum WebserverFlutterRouteAuth { publicAccess, requiredAccess }

/// Navigation slot a route occupies, and its idle position in the tab bar.
class WebserverFlutterNavigationContribution {
  const WebserverFlutterNavigationContribution({
    required this.labelKey,
    required this.permission,
    required this.order,
  });

  final String labelKey;
  final String permission;
  final int order;
}

class WebserverFlutterRouteContribution {
  const WebserverFlutterRouteContribution({
    required this.id,
    required this.domain,
    required this.capability,
    required this.screen,
    required this.routeName,
    required this.titleKey,
    required this.auth,
    this.surface = 'app',
    this.permissionHint,
    this.navigation,
  });

  final String id;
  final String surface;
  final String domain;
  final String capability;
  final String screen;

  /// Physical Flutter route name, e.g. `/applications`.
  final String routeName;

  final String titleKey;
  final WebserverFlutterRouteAuth auth;

  /// Permission the owning module derives from the route's own operationId.
  final String? permissionHint;

  final WebserverFlutterNavigationContribution? navigation;
}

/// Validate the canonical route id shape, uniqueness, and the
/// placement/identity agreement.
///
/// Returns issues rather than throwing so verification can report every
/// violation in one run.
List<String> validateWebserverFlutterRouteContributions(
  List<WebserverFlutterRouteContribution> routes,
) {
  final issues = <String>[];
  final seenIds = <String>[];
  final seenRouteNames = <String>[];
  for (final route in routes) {
    final expectedId =
        '${route.surface}.${route.domain}.${route.capability}.${route.screen}';
    if (route.id != expectedId) {
      issues.add('route id ${route.id} must equal $expectedId');
    }
    if (seenIds.contains(route.id)) {
      issues.add('duplicate flutter route id ${route.id}');
    }
    seenIds.add(route.id);
    if (route.routeName.trim().isEmpty) {
      issues.add('route ${route.id} must declare a route name');
    }
    if (seenRouteNames.contains(route.routeName)) {
      issues.add('duplicate flutter route name ${route.routeName}');
    }
    seenRouteNames.add(route.routeName);
    if (route.navigation != null && route.permissionHint == null) {
      issues.add(
        'route ${route.id} contributes navigation but declares no permissionHint',
      );
    }
  }
  return issues;
}

/// Assemble the router table from capability contributions.
///
/// Contributions with no route name are dropped rather than mounted at an empty
/// path, which would swallow every other route.
List<WebserverFlutterRouteContribution> createWebserverFlutterRouteRegistry(
  List<WebserverFlutterRouteContribution> routes,
) {
  return routes
      .where((route) => route.routeName.trim().isNotEmpty)
      .toList(growable: false);
}

/// First route the shell mounts, preferring the lowest navigation order.
String? resolveWebserverFlutterHomeRouteName(
  List<WebserverFlutterRouteContribution> routes,
) {
  final navigable =
      routes.where((route) => route.navigation != null).toList(growable: false)
        ..sort((a, b) => a.navigation!.order.compareTo(b.navigation!.order));
  if (navigable.isNotEmpty) {
    return navigable.first.routeName;
  }
  return routes.isEmpty ? null : routes.first.routeName;
}

File: apps/sdkwork-webserver-h5/packages/sdkwork-webserver-h5-shell/src/navigation/routeRegistry.ts
/**
 * Web Server H5 navigation model.
 *
 * Route ownership lives in the shell so the application root only assembles
 * entries into a router and a feature package never has to know the mobile URL
 * it happens to live under. Physical H5 paths stay short (`/applications`)
 * while `id` carries the canonical cross-client identity
 * (`APP_CLIENT_ARCHITECTURE_ALIGNMENT_SPEC.md` section 7):
 *
 * ```text
 * <surface>.<domain>.<capability>.<screen>
 * ```
 *
 * Section 7: "SDKWork client routes are aligned by route identity, not by
 * identical physical paths", and the `id`, `surface`, `domain`, `capability`,
 * `screen`, `titleKey`, and `permissionHint` tuple must stay consistent across
 * PC, H5, Flutter, mini program, Android, iOS, and Harmony implementations of
 * one workflow.
 *
 * History: this module previously used the bare capability segment
 * (`applications`) as the route `id`, citing the cross-client capability
 * segment. Section 7 owns route identity and mandates four segments, so the id
 * was migrated and `createWebserverH5Navigation` now refuses a non-canonical id
 * rather than letting the drift reappear.
 */

/** Cross-client surface segment for every application-surface H5 route. */
export const WEBSERVER_H5_ROUTE_SURFACE = "app" as const;

/** Cross-client domain segment for Web Server routes. */
export const WEBSERVER_H5_ROUTE_DOMAIN = "webserver" as const;

/**
 * Cross-client capability segment of the applications workflow.
 *
 * It matches the PC console resource key and the `deploy.apps.read` permission
 * resource, so a navigation entry and its permission hint never drift apart.
 * The physical H5 path is derived from it; the route **id** is not.
 */
export const APPLICATIONS_ROUTE = "applications" as const;

/** Cross-client screen segment of the applications list workflow. */
export const APPLICATIONS_SCREEN = "list" as const;

/** Canonical cross-client route id of the applications list workflow. */
export const APPLICATIONS_ROUTE_ID =
  `${WEBSERVER_H5_ROUTE_SURFACE}.${WEBSERVER_H5_ROUTE_DOMAIN}.${APPLICATIONS_ROUTE}.${APPLICATIONS_SCREEN}` as const;

/** Mobile path of the applications surface. */
export const APPLICATIONS_PATH = `/${APPLICATIONS_ROUTE}` as const;

export interface WebserverH5NavigationEntry {
  /** Canonical route id: `<surface>.<domain>.<capability>.<screen>`. */
  readonly id: string;
  /** Cross-client surface segment; identical on every client surface. */
  readonly surface: string;
  /** Cross-client domain segment. */
  readonly domain: string;
  /** Cross-client capability segment. */
  readonly capability: string;
  /** Cross-client screen segment. */
  readonly screen: string;
  /** Absolute H5 path the application root registers in its router. */
  readonly path: string;
  /** Catalog key of the tab-bar label, never inline copy. */
  readonly labelKey: string;
  /** Catalog key of the screen title, shared with the other client roots. */
  readonly titleKey: string;
  /**
   * Permission required to see the entry, using the same `[resource].[action]`
   * shape the gateway derives from `operationId`. The shell only carries the
   * hint; the root decides whether the viewer holds it.
   */
  readonly permissionHint: string;
  /** Ascending display order in the tab bar. */
  readonly order: number;
}

export function validateWebserverH5NavigationEntries(
  entries: readonly WebserverH5NavigationEntry[],
): readonly string[] {
  const issues: string[] = [];
  for (const entry of entries) {
    const expectedId = `${entry.surface}.${entry.domain}.${entry.capability}.${entry.screen}`;
    if (entry.id !== expectedId) {
      issues.push(`route id ${entry.id} must equal ${expectedId}`);
    }
    if (entry.path.trim().length === 0) {
      issues.push(`route ${entry.id} must declare a path`);
    }
    if (entry.titleKey.trim().length === 0) {
      issues.push(`route ${entry.id} must declare a titleKey`);
    }
    if (entry.permissionHint.trim().length === 0) {
      issues.push(`route ${entry.id} must declare a permissionHint`);
    }
  }
  return issues;
}

/**
 * Entries the H5 shell itself owns: chrome-level destinations that are not the
 * product of any single feature package.
 */
export const WEBSERVER_H5_SHELL_NAVIGATION: readonly WebserverH5NavigationEntry[] = [];

/**
 * The applications entry contributed by `@sdkwork/webserver-h5-applications`.
 *
 * It is declared here rather than inside the feature package so the shell stays
 * the single place that knows the ordered navigation model, and the feature
 * package stays the single place that knows how the screen is rendered.
 */
export const WEBSERVER_H5_APPLICATIONS_NAVIGATION: readonly WebserverH5NavigationEntry[] = [
  {
    id: APPLICATIONS_ROUTE_ID,
    surface: WEBSERVER_H5_ROUTE_SURFACE,
    domain: WEBSERVER_H5_ROUTE_DOMAIN,
    capability: APPLICATIONS_ROUTE,
    screen: APPLICATIONS_SCREEN,
    path: APPLICATIONS_PATH,
    labelKey: "navigation.applications",
    titleKey: "applications.list.title",
    permissionHint: "deploy.apps.read",
    order: 10,
  },
];

/** Path the router falls back to when nothing matched. */
export const WEBSERVER_H5_HOME_PATH = APPLICATIONS_PATH;

export function createWebserverH5Navigation(
  contributions: readonly (readonly WebserverH5NavigationEntry[])[],
): readonly WebserverH5NavigationEntry[] {
  const merged = contributions.flat();

  const issues = validateWebserverH5NavigationEntries(merged);
  if (issues.length > 0) {
    throw new Error(issues.join("; "));
  }

  const seenIds = new Set<string>();
  const seenPaths = new Set<string>();
  for (const entry of merged) {
    if (seenIds.has(entry.id)) {
      throw new Error(`duplicate H5 navigation id: ${entry.id}`);
    }
    if (seenPaths.has(entry.path)) {
      throw new Error(`duplicate H5 navigation path: ${entry.path}`);
    }
    seenIds.add(entry.id);
    seenPaths.add(entry.path);
  }
  return [...merged].sort((left, right) => left.order - right.order);
}

/** Entries the viewer may open, in tab-bar order. */
export function filterNavigationForPermission(
  entries: readonly WebserverH5NavigationEntry[],
  hasPermission: (permission: string) => boolean,
): readonly WebserverH5NavigationEntry[] {
  return entries.filter((entry) => hasPermission(entry.permissionHint));
}
